pageserver/src/tenant/storage_layer/inmemory_layer.rs

```diff
@@ -456,44 +620,66 @@ impl InMemoryLayer {
         })
     }
 
-    // Write path.
+    /// Write path.
+    ///
+    /// Errors are not retryable, the [`InMemoryLayer`] must be discarded, and not be read from.
+    /// The reason why it's not retryable is that the [`EphemeralFile`] writes are not retryable.
+    /// TODO: it can be made retryable if we aborted the process on EphemeralFile write errors.
     pub async fn put_batch(
         &self,
         serialized_batch: SerializedBatch,
         ctx: &RequestContext,
-    ) -> Result<()> {
+    ) -> anyhow::Result<()> {
         let mut inner = self.inner.write().await;
         self.assert_writable();
 
-        let base_off = {
-            inner
-                .file
-                .write_raw(
-                    &serialized_batch.raw,
-                    &RequestContextBuilder::extend(ctx)
-                        .page_content_kind(PageContentKind::InMemoryLayer)
-                        .build(),
-                )
-                .await?
-        };
-
-        for SerializedBatchOffset {
-            key,
-            lsn,
-            offset: relative_off,
-        } in serialized_batch.offsets
-        {
-            let off = base_off + relative_off;
+        let base_offset = inner.file.len();
+
+        // Add the base_offset to the batch's index values which are relative to the batch start.
+        let index_values: Vec<SerializedBatchOffset> = serialized_batch
+            .offsets
+            .into_iter()
+            .map(|SerializedBatchOffset { key, lsn, value }| {
+                let InMemoryLayerIndexValueUnpacked {
+                    will_init,
+                    len,
+                    pos,
+                } = value.unpack();
+                anyhow::Ok(SerializedBatchOffset {
+                    key,
+                    lsn,
+                    value: InMemoryLayerIndexValue::new(InMemoryLayerIndexValueNewArgs {
+                        base_offset,
+                        batch_offset: pos,
+                        len: len.into_usize(),
+                        will_init,
+                    })?,
```

Comment:
Can we make the `?` fallibility here into an assertion?  There's no nice way to handle this error, if we've hit some bug that has us writing values too big to pack.

Reply:
It was an `expect()` before. @VladLazar  suggested to change it into an anyhow error.

I think anyhow makes more sense because once we separate the producer of the SerializedBatch from the consumer (i.e., producer will be SK and consumer will be PS), we should not panic on unexpected input.